src/libstd/sys/windows/fs.rs

```diff
@@ -567,23 +646,85 @@ pub fn remove_dir_all(path: &Path) -> io::Result<()> {
+    // instantaneously, but is scheduled. So just before or after flagging a
+    // file for deletion, we move it to some `base_dir` to avoid races.
+
+    fn move_item(file: &File, base_dir: &Path, mut counter: u64)
+            -> io::Result<u64> {
+        let mut tmpname = base_dir.join(format!{"rm-{}", counter});
+        counter += 1;
+        // Try to rename the file. If it already exists, just retry with an other filename.
+        while let Err(err) = file.rename(tmpname.as_ref(), false) {
+            if err.kind() != io::ErrorKind::AlreadyExists { return Err(err) };
+            tmpname = base_dir.join(format!{"rm-{}", counter});
+            counter += 1;
+        }
+        Ok(counter)
+    }
+
+    fn remove_item(path: &Path,
+                   base_dir: &Path,
+                   mut counter: u64,
+                   readonly: bool) -> io::Result<u64> {
+        if !readonly {
+            let mut opts = OpenOptions::new();
+            opts.access_mode(c::DELETE);
+            opts.custom_flags(c::FILE_FLAG_BACKUP_SEMANTICS | // delete directory
+                              c::FILE_FLAG_OPEN_REPARSE_POINT | // delete symlink
+                              c::FILE_FLAG_DELETE_ON_CLOSE);
+            let file = try!(File::open(path, &opts));
+            counter = try!(move_item(&file, base_dir, counter));
+        } else {
+            let mut opts = OpenOptions::new();
+            opts.access_mode(c::DELETE | c::FILE_WRITE_ATTRIBUTES);
+            opts.custom_flags(c::FILE_FLAG_OPEN_REPARSE_POINT);
+            let file = try!(File::open(path, &opts));
+            // remove read-only permision
+            try!(file.set_perm(FilePermissions::new()));
+            counter = try!(move_item(&file, base_dir, counter));
+            try!(file.remove());
+            // restore read-only flag just in case there are other hard links
+            let mut perm = FilePermissions::new();
+            perm.set_readonly(true);
+            try!(file.set_perm(perm));
+        };
+        Ok(counter)
+    }
+
     for child in try!(readdir(path)) {
         let child = try!(child);
         let child_type = try!(child.file_type());
         if child_type.is_dir() {
-            try!(remove_dir_all_recursive(&child.path()));
-        } else if child_type.is_symlink_dir() {
-            try!(rmdir(&child.path()));
+            counter = try!(remove_dir_all_recursive(&child.path(),
+                                                    base_dir,
+                                                    counter));
         } else {
-            try!(unlink(&child.path()));
+            counter = try!(remove_item(&child.path().as_ref(),
+                                       base_dir,
+                                       counter,
+                                       try!(child.metadata()).perm().readonly()));
         }
     }
-    rmdir(path)
+    counter = try!(remove_item(path, base_dir, counter, false));
```

Comment:
Managing this `counter` variable may be best done with some sort of `Context` structure or something like that so it doesn't have to constantly be passed in and reset.